vulnerabilities/importers/rust.py

```diff
@@ -1,99 +1,176 @@
+            resolved_package_urls=[PackageURL(type='cargo', name=crate_name, version=v) for v in unaffected],
+            reference_urls=[reference_url] if reference_url else [],
+            reference_ids=[advisory['id']],
+            cve_id=cve_id,
+        )
+
+
+RUSTSEC_DB_URL = 'https://github.com/RustSec/advisory-db/archive/master.zip'
 
-    elif not have_affected_version_range:
-        affected_versions = all_versions - unaffected_versions
 
-    return {
-        'package_name': crate_name,
-        'vuln_id': vuln_id,
-        'fixed_versions': unaffected_versions,
-        'affected_versions': affected_versions,
-        'advisory': reference,
-        'description': description
-    }
+def rust_crate_advisories(url, prefix='advisory-db-master/crates/'):
+    with urlopen(url) as response:
+        with ZipFile(BytesIO(response.read())) as zf:
+            for path in zf.namelist():
+                if path.startswith(prefix) and path.endswith('.toml'):
+                    yield toml.load(zf.open(path))
 
 
-def import_vulnerabilities():
+def categorize_versions(
+        all_versions: Set[str],
+        unaffected_versions: Set[RangeSpecifier],
+        affected_versions: Set[RangeSpecifier],
+        resolved_versions: Set[RangeSpecifier],
+) -> Tuple[Set[str], Set[str]]:
+    """
+    Categorize all versions of a crate according to the given version ranges.
 
-    vulnerability_package_dicts = []
-    for advisory in rust_crate_advisories(RUSTSEC_DB_URL):
+    :return: unaffected versions, affected versions
+    """
+    unaffected, affected = set(), set()
 
-        loaded_advisory = load_advisory(advisory)
-        if loaded_advisory:
-            vulnerability_package_dicts.append(loaded_advisory)
+    if not any(unaffected_versions.union(affected_versions).union(resolved_versions)):
+        return unaffected, affected
 
-    return vulnerability_package_dicts
+    for version in all_versions:
+        if affected_versions and all([version in av for av in affected_versions]):
+            affected.add(version)
+        elif unaffected_versions and all([version in av for av in unaffected_versions]):
+            unaffected.add(version)
+        elif resolved_versions and all([version in av for av in resolved_versions]):
+            unaffected.add(version)
+
+    # If some versions were not classified above, one or more of the given ranges might be empty, so the remaining
+    # versions default to either affected or unaffected.
+    uncategorized_versions = all_versions - unaffected.union(affected)
+    if uncategorized_versions:
+        if not affected_versions:
+            affected.update(uncategorized_versions)
+        else:
+            unaffected.update(uncategorized_versions)
+
+    return unaffected, affected
+
+
+class CratesAPI:
+    def __init__(self, cache=None):
+        self.versions_cache = cache or {}
+
+    def all_versions_of_crate(self, crate: str) -> Set[str]:
+        if crate in self.versions_cache:
```

Comment:
I haven't finished reviewing this in detail. So I might be missing some things here . 
This API provides last modification date, how about using it , to update the cache.

I don't know what difference it would make, but will using a single  `requests.Session()`  instead of the current solution to make  all the API calls increase the performance,  https://2.python-requests.org/en/v2.8.1/user/advanced/#session-objects ?
Actually we could make huge gains in performance at the cost of increased complexity using asynchronous API calls, we won't dive into that.  

Btw @pombredanne  mentioned in the chat about having  some code for obtaining data from APIs provided by  package managers.